Python/Product/PythonTools/PythonTools/Environments/AddVirtualEnvironmentOperation.cs

```diff
@@ -0,0 +1,183 @@
+        private readonly bool _installReqs;
+        private readonly string _reqsPath;
+        private readonly bool _registerAsCustomEnv;
+        private readonly string _customEnvName;
+        private readonly bool _setAsCurrent;
+        private readonly bool _setAsDefault;
+        private readonly bool _viewInEnvWindow;
+        private readonly Redirector _output;
+        private readonly IVsTaskStatusCenterService _statusCenter;
+        private readonly IInterpreterRegistryService _registry;
+        private readonly IInterpreterOptionsService _options;
+        private readonly IPythonToolsLogger _logger;
+
+        public AddVirtualEnvironmentOperation(
+            IServiceProvider site,
+            PythonProjectNode project,
+            string virtualEnvPath,
+            string baseInterpreterId,
+            bool useVEnv,
+            bool installRequirements,
+            string requirementsPath,
+            bool registerAsCustomEnv,
+            string customEnvName,
+            bool setAsCurrent,
+            bool setAsDefault,
+            bool viewInEnvWindow,
+            Redirector output = null
+        ) {
+            _site = site ?? throw new ArgumentNullException(nameof(site));
+            _project = project;
+            _virtualEnvPath = virtualEnvPath ?? throw new ArgumentNullException(nameof(virtualEnvPath));
+            _baseInterpreter = baseInterpreterId ?? throw new ArgumentNullException(nameof(baseInterpreterId));
+            _useVEnv = useVEnv;
+            _installReqs = installRequirements;
+            _reqsPath = requirementsPath;
+            _registerAsCustomEnv = registerAsCustomEnv;
+            _customEnvName = customEnvName;
+            _setAsCurrent = setAsCurrent;
+            _setAsDefault = setAsDefault;
+            _viewInEnvWindow = viewInEnvWindow;
+            _output = output;
+            _statusCenter = _site.GetService(typeof(SVsTaskStatusCenterService)) as IVsTaskStatusCenterService;
+            _registry = _site.GetComponentModel().GetService<IInterpreterRegistryService>();
+            _options = _site.GetComponentModel().GetService<IInterpreterOptionsService>();
+            _logger = _site.GetService(typeof(IPythonToolsLogger)) as IPythonToolsLogger;
+        }
+
+        public async Task RunAsync() {
+            var outputWindow = OutputWindowRedirector.GetGeneral(_site);
+            var taskHandler = _statusCenter.PreRegister(
+                new TaskHandlerOptions() {
+                    ActionsAfterCompletion = CompletionActions.RetainAndNotifyOnFaulted | CompletionActions.RetainAndNotifyOnRanToCompletion,
+                    Title = Strings.VirtualEnvStatusCenterCreateTitle.FormatUI(PathUtils.GetFileOrDirectoryName(_virtualEnvPath)),
+                    DisplayTaskDetails = (t) => { outputWindow.ShowAndActivate(); }
+                },
+                new TaskProgressData() {
+                    CanBeCanceled = false,
```

Comment:
So why cancellation isn't allowed here?

Reply:
We never allowed cancel of this operation in the past.
Killing the process while it's creating/installing will end up with an unusable env on disk.

We could cancel between creation and installation though, since that's 2 consecutive processes that are launched. But sometimes there's no installation step.

So giving them a cancel button that does pretty much nothing may just lead to more frustration?

I'll put it on the list to consider.